url78d0cbe382_TeeAaTeeUu_SitsiPlaseeraus_tgz-pJ8-sitsiplaseeraus_MainJ8/src/sitsiplaseeraus/Sitsaaja.java

```diff
@@ -0,0 +1,249 @@
+     */
+    public void setPuoliso(Sitsaaja puoliso) {
+        this.puoliso = puoliso;
+        this.setPuolisoIsSet();
+    }
+
+    /**
+     * palauttaa paikan jossa istuu, jos kerrottu
+     *
+     * @return Paikka luokka
+     */
+    public Paikka getPaikka() {
+        return paikka;
+    }
+
+    /**
+     * Palauttaa paikan, jossa luulee istuvansa
+     *
+     * @param paikka Paikka luokka
+     */
+    public void setPaikka(Paikka paikka) {
+        this.paikka = paikka;
+    }
+
+    /**
+     * onko sitsaajalla avecia
+     *
+     * @return true tai false
+     */
+    public boolean avecIsSet() {
+        return avecIsSet;
+    }
+
+    private void setAvecIsSet() {
+        this.avecIsSet = true;
+    }
+
+    /**
+     * onko sitsaajalla puolisoa
+     *
+     * @return true tai false
+     */
+    public boolean puolisoIsSet() {
+        return puolisoIsSet;
+    }
+
+    private void setPuolisoIsSet() {
+        this.puolisoIsSet = true;
+    }
+
+    /**
+     * Asettaa yhteyden sitsaajalle, eli kenestä toisesta sitsaajasta pitää tai
+     * vihaa
+     *
+     * @param sitsaaja
+     * @param arvo kuinka paljon vihaa tai rakastaa, väliltä -5 ja 5
+     * @return jos arvo oli huono niin false, muuten true
+     */
+    public boolean setYhteys(Sitsaaja sitsaaja, int arvo) {
+        if (this.equals(sitsaaja) || arvo > 5 || arvo < -5 || this.yhteydet.containsKey(sitsaaja)) {
+            return false;
+        }
+        this.yhteydet.put(sitsaaja, arvo);
+        return true;
+    }
+
+    /**
+     * Poistaa yhteyden sitsaajalta, jos sellainen on olemassa
+     *
+     * @param sitsaaja
+     * @return onnistuiko vai ei
+     */
+    public boolean deleteYhteys(Sitsaaja sitsaaja) {
+        if (this.yhteydet.containsKey(sitsaaja)) {
+            this.yhteydet.remove(sitsaaja);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Palauttaa sitsaajan kaikki yhteydet
+     *
+     * @return yhteydet
+     */
+    public Hakemisto<Sitsaaja, Integer> palautaYhteydet() {
+        return this.yhteydet;
+    }
+
+    /**
+     * Kuinka paljon sitsaajalla on yhteyksiä muihin sitsaajiin
+     *
+     * @return määrä
+     */
+    public int yhteyksienMaara() {
+        int yhteyksienMaara = 0;
+        for (Vektori<Sitsaaja, Integer> yhteys : this.yhteydet) {
+            yhteyksienMaara++;
+        }
+        return yhteyksienMaara;
+    }
+
+    public boolean heteroPari() {
+        if (avecIsSet == false && puolisoIsSet == false) {
+            return false;
+        }
+        if (avecIsSet) {
+            if (isMies()) {
+                if (getAvec().isMies()) {
+                    return false;
+                } else {
+                    return true;
+                }
+            } else {
+                if (getAvec().isMies()) {
+                    return true;
+                } else {
+                    return false;
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 220:232, 233:247.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.